Increase Nonbonded threads on AMD GPUs

bdenhollander opened the pull request:
I've mentioned this optimization before for [RDNA](https://github.com/openmm/openmm/issues/3937#issuecomment-1416579982) and in the past for [GCN](https://github.com/openmm/openmm/issues/2894). This PR limits the impact of this change to AMD GPUs and avoids spawning unnecessary threads on small systems.

Graphs from AMD GPU Profiler showing the increase in reported utilization are available [here](https://github.com/openmm/openmm/pull/3911#issuecomment-1386414363). The code to reduce thread blocks for small systems avoids the problematic gbsa 8x and 16x scenarios included in the previous link.

This change should work well on [Radeon RX 5700 XT](https://github.com/openmm/openmm/issues/2894#issuecomment-716795688) and 16x is already being used in the [Metal plugin](https://github.com/philipturner/openmm-metal/blob/6911d157f3aa1a9263aa53faf68fefa7a33fd3d9/platforms/metal/src/MetalNonbondedUtilities.cpp#L74-L77).

```
benchmark.py --precision single --platform OpenCL --seconds 30 --verbose --style table
```
Radeon RX 6600 benchmarks on Windows 10

|            | Original<br/>ns/day | 16x<br/>ns/day | Change |
|------------|-------------|---------|--------|
| gbsa       | 411.830     | 411.122 | -0.17% |
| rf         | 437.085     | 456.208 | 4.38%  |
| pme        | 304.292     | 327.070 | 7.49%  |
| apoa1rf    | 112.201     | 123.168 | 9.77%  |
| apoa1pme   | 79.6377     | 85.3634 | 7.19%  |
| apoa1ljpme | 62.3718     | 65.6956 | 5.33%  |
| amoebagk   | 1.41290     | 1.41358 | 0.05%  |
| amoebapme  | 3.96598     | 4.04453 | 1.98%  |

gbsa doesn't have enough atoms to benefit from the change and both amoeba tests spend a lot of more of their time on kernels other than `computeNonbonded`.

| FAH Project | Particles | Original<br/>ns/day | x16<br/>ns/day | Change |
|-------------|-----------|-----------------|------------|--------|
| 12000       | 59495     | 66.0092         | 67.287     | 1.94%  |
| 13415       | 4038      | 127.6969        | 132.3496   | 3.64%  |
| 13428       | 90729     | 10.1501         | 10.4697    | 3.15%  |
| 14908       | 215357    | 17.8778         | 19.1612    | 7.18%  |
| 14944       | 105522    | 41.8681         | 42.6525    | 1.87%  |
| 16918       | 65523     | 49.6566         | 53.0001    | 6.73%  |
| 17308       | 158384    | 46.3842         | 50.6529    | 9.20%  |
| 17424       | 167028    | 47.1297         | 50.4466    | 7.04%  |
| 17600       | 37801     | 132.1982        | 133.852    | 1.25%  |
| 17919       | 103824    | 41.3429         | 42.1834    | 2.03%  |
| 18213       | 302725    | 13.5422         | 14.322     | 5.76%  |
| 18449       | 142541    | 26.4122         | 28.3827    | 7.46%  |
| 18717       | 77795     | 36.08           | 38.3604    | 6.32%  |

bdenhollander committed:
Increase Nonbonded threads on AMD

Use 4x more threads on AMD GPUs to increase utilization. Reduce the number of threads if simulation is very small to avoid spawning threads that will do no work.

bdenhollander commented:
@ex-rzr Are you be able to test this change on your fleet of AMD GPUs?

peastman commented:
Here's what I get for RX 5700 XT on Linux.

|Test|Old|New|Ratio|
|---|---|---|---|
gbsa|461.329|363.43|0.79
rf|561.201|617.506|1.10
pme|357.602|343.165|0.96
apoa1rf|151.315|164.478|1.09
apoa1pme|104.327|112.568|1.08
apoa1ljpme|84.0918|86.0825|1.02
amoebagk|1.06799|1.31854|1.23
amoebapme|5.54152|5.48906|0.99
amber20-dhfr|412.945|363.256|0.88
amber20-cellulose|20.8944|21.873|1.05
amber20-stmv|5.09852|5.24142|1.03

There's no consistent benefit.  It's slower on some, faster on others.

peastman commented:
It reports 20.

> The following loop condition would reduce threads less.

Won't that reduce the threads more, not less?

bdenhollander commented:
> Won't that reduce the threads more, not less?

Whoops, wrong operator. Fixed in the most recent edit.

peastman commented:
Here are the numbers with that change.

|Test|Old|New|Ratio|
|---|---|---|---|
gbsa|461.329|409.76|0.89
rf|561.201|608.765|1.08
pme|357.602|356.017|1.00
apoa1rf|151.315|164.437|1.09
apoa1pme|104.327|112.695|1.08
apoa1ljpme|84.0918|86.7135|1.03
amoebagk|1.06799|1.04627|0.98
amoebapme|5.54152|5.31066|0.96
amber20-dhfr|412.945|418.811|1.01
amber20-cellulose|20.8944|21.8759|1.05
amber20-stmv|5.09852|5.2052|1.02

bdenhollander commented:
I attempted to reproduce the gbsa performance regression on my RX 6600 on Ubuntu 20.04.06 but I'm not seeing 1-3% rather than 10-20%. Below are results for v1 of the thread reduction algorithm, v2 with the `/2` in the the `where`, and then without the thread reduction for low atom counts.

The goal of the thread reduction was to avoid regressing gbsa but maybe it's not necessary. I will need to do another run on Windows to see how much of a difference it made there.

|            | Original | 16x Reduce<br>Threads v1 | Ratio | 16x Reduce<br>Threads v2 | Ratio | 16x No Reduce<br>Threads | Ratio |
|------------|----------|---------------|-------|---------------|-------|---------------|-------|
| gbsa       | 525.632  | 518.835       | 0.99  | 507.339       | 0.97  | 514.513       | 0.98  |
| rf         | 563.677  | 595.621       | 1.06  | 600.737       | 1.07  | 601.537       | 1.07  |
| pme        | 292.947  | 304.698       | 1.04  | 302.689       | 1.03  | 302.546       | 1.03  |
| apoa1rf    | 132.948  | 146.236       | 1.10  | 145.805       | 1.10  | 145.448       | 1.09  |
| apoa1pme   | 79.3241  | 83.0128       | 1.05  | 83.233        | 1.05  | 82.9484       | 1.05  |
| apoa1ljpme | 60.6558  | 63.4862       | 1.05  | 63.3956       | 1.05  | 63.3767       | 1.04  |
| amoebagk   | 1.01827  | 1.01907       | 1.00  | 1.08331       | 1.06  | 1.17638       | 1.16  |
| amoebapme  | 3.85564  | 3.89434       | 1.01  | 3.837         | 1.00  | 3.83167       | 0.99  |

bdenhollander commented:
RX 6600 rerun on Windows 10.

|            | Original | 16x Reduce Threads v1 | Ratio | 16x Reduce Threads v2 | Ratio | 16x No Reduce Threads | Ratio |
|------------|----------|-----------------------|-------|-----------------------|-------|-----------------------|-------|
| gbsa       | 411.83   | 411.122  | 1.00  | 401.912 | 0.98  | 407.451| 0.99  |
| rf         | 437.085  | 456.208  | 1.04  | 472.745 | 1.08  | 459.72 | 1.05  |
| pme        | 304.292  | 327.07 | 1.07  | 333.446 | 1.10  | 325.636 | 1.07  |
| apoa1rf    | 112.201  | 123.168  | 1.10  | 118.341| 1.05  | 124.64 | 1.11  |
| apoa1pme   | 79.6377  | 85.3634  | 1.07  | 83.2742 | 1.05  | 82.691 | 1.04  |
| apoa1ljpme | 62.3718  | 65.6956 | 1.05  | 65.615 | 1.05  | 65.0471 | 1.04  |
| amoebagk   | 1.4129   | 1.41358 | 1.00  | 1.27443 | 0.90  | 1.26674 | 0.90  |
| amoebapme  | 3.96598  | 4.04453 | 1.02  | 4.0343 | 1.02  | 3.97247 | 1.00  |

Reducing threads for low atom counts avoids a minor 1% regression in gbsa on Windows. Interestingly, amoebagk moves in the opposite directions when not adjusting for atom count on Windows, -10% while Linux is +16%.

bdenhollander committed:
Merge remote-tracking branch 'upstream/master' into 16x-nonbonded

bdenhollander committed:
Merge remote-tracking branch 'upstream/master' into 16x-nonbonded

bdenhollander committed:
Use isAMD

bdenhollander committed:
RDNA2 tuning

- Use 16 as the multiplier in OpenCLContext and OpenCLNonbonded Utilities
- Add getReducedNumThreadBlocks to prevent performance regression on amoebagk and amoebapme

bdenhollander committed:
Merge remote-tracking branch 'upstream/master' into 16x-nonbonded

bdenhollander commented:
Updated benchmarks for OpenMM 8.1 on RX 6000 OpenCL 3592.0 on Windows.

| Benchmark         | OpenMM 8.1<br/>ns/day | Tuned<br/>ns/day | Ratio |
|-------------------|-------------------|--------------|-------|
| gbsa              | 655.36            | 655.675      | 1.000 |
| rf                | 570.619           | 612.108      | 1.073 |
| pme               | 439.244           | 457.137      | 1.041 |
| apoa1rf           | 161.246           | 176.841      | 1.097 |
| apoa1pme          | 107.543           | 117.403      | 1.092 |
| apoa1ljpme        | 81.799            | 89.2548      | 1.091 |
| amoebagk          | 1.3965            | 1.37489      | 0.985 |
| amoebapme         | 4.99101           | 5.64573      | 1.131 |
| amber20-dhfr      | 478.048           | 500.196      | 1.046 |
| amber20-cellulose | 24.8385           | 27.0305      | 1.088 |
| amber20-stmv      | 7.18744           | 7.92239      | 1.102 |

Water boxes

| Width | Atoms   | OpenMM 8.1<br/>ns/day | Tuned<br/>ns/day | Ratio |
|-------|---------|-----------------|--------------|-------|
| 4     | 6282    | 853             | 863          | 1.012 |
| 5     | 12255   | 574             | 603          | 1.051 |
| 6     | 21384   | 467             | 492          | 1.054 |
| 7     | 33840   | 315             | 333          | 1.057 |
| 8     | 50442   | 213             | 226          | 1.061 |
| 9     | 72279   | 147             | 160          | 1.088 |
| 10    | 98880   | 107             | 115          | 1.075 |
| 11    | 131445  | 79.2            | 87.6         | 1.106 |
| 15    | 335025  | 29.9            | 32.9         | 1.100 |
| 20    | 792450  | 11.7            | 13.1         | 1.120 |
| 25    | 1550160 | 5.25            | 5.79         | 1.103 |

FAH Projects

| Project | Atoms  | OpenMM 8.1<br/>ns/day | Tuned<br/>ns/day | Ratio |
|---------|--------|---------------------|--------------|-------|
| 12271   | 42065  | 201.571             | 217.9836     | 1.081 |
| 13415   | 4038   | 174.5842            | 175.1544     | 1.003 |
| 13428   | 90729  | 13.3166             | 14.2577      | 1.071 |
| 14908   | 215357 | 17.6863             | 19.8135      | 1.120 |
| 14944   | 105522 | 41.0235             | 42.8305      | 1.044 |
| 16918   | 65523  | 69.0273             | 75.3143      | 1.091 |
| 17308   | 158384 | 40.9228             | 45.0127      | 1.100 |
| 17424   | 167028 | 54.8505             | 59.5879      | 1.086 |
| 17600   | 37801  | 199.2321            | 209.0226     | 1.049 |
| 17919   | 103824 | 46.5742             | 48.3243      | 1.038 |
| 18213   | 302725 | 15.2943             | 16.3364      | 1.068 |
| 18449   | 142541 | 29.1367             | 32.461       | 1.114 |
| 18717   | 77795  | 50.1016             | 53.4198      | 1.066 |

bdenhollander committed:
Add getReducedNumThreadBlocks to CUDA

bdenhollander committed:
Move reduced count from ComputeContext to NonbondedUtilities

- Attempt to fix AmoebaGK Jenkins failures since CUDA thread blocks is different between context (x4/x6) and nonbonded utilities (x4)

bdenhollander commented:
@peastman Is there any chance you could benchmark this branch again on your wider 5700 XT?
